### PR TITLE
feat: pass `access_token` to `coder_git_auth` resource

### PR DESCRIPTION
**Pull request by kylecarbs**
This allows template authors to leverage git auth to perform custom actions, like clone repositories.

I mostly moved code around in this, there are actually very few changes. The increased lines is primarily due to hardening around refreshing the Git token, which I felt was important enough to become slightly abstracted.

This essentially enables:

```tf
data "coder_git_auth" "github" {
  id = "github"
}

resource "coder_agent" "dev" {
  env = {
    GITHUB_TOKEN: data.coder_git_auth.github.access_token,
  }
}
```

### CODE REVIEW

File: coderd/provisionerdserver/provisionerdserver.go [hunk 1]
```diff
@@ -210,6 +210,44 @@ func (server *Server) AcquireJob(ctx context.Context, _ *proto.Empty) (*proto.Ac
...
 		}
 
+		gitAuthProviders := []*sdkproto.GitAuthProvider{}
+		for _, p := range templateVersion.GitAuthProviders {
+			link, err := server.Database.GetGitAuthLink(ctx, database.GetGitAuthLinkParams{
+				ProviderID: p,
+				UserID:     owner.ID,
+			})
+			if errors.Is(err, sql.ErrNoRows) {
+				continue
```
Comment (deansheather) [inline coderd/provisionerdserver/provisionerdserver.go:220]: I assumed that git auth was required if a template specifies it based on not being able to create a workspace in dogfood until I set it up. Should this not fail the build early or something?

File: coderd/provisionerdserver/provisionerdserver.go [hunk 2]
```diff
@@ -210,6 +210,44 @@ func (server *Server) AcquireJob(ctx context.Context, _ *proto.Empty) (*proto.Ac
...
+			for _, c := range server.GitAuthConfigs {
+				if c.ID != p {
+					continue
+				}
+				config = c
+				break
+			}
+			// We weren't able to find a matching config for the ID!
+			if config == nil {
+				continue
```
Comment (deansheather) [inline coderd/provisionerdserver/provisionerdserver.go:235]: We should probably log and/or fail the build here.

File: provisionerd/proto/provisionerd.proto [hunk 3]
```diff
@@ -17,8 +17,9 @@ message AcquiredJob {
         repeated provisioner.ParameterValue parameter_values = 3;
         repeated provisioner.RichParameterValue rich_parameter_values = 4;
 		repeated provisioner.VariableValue variable_values = 5;
-        provisioner.Provision.Metadata metadata = 6;
-        bytes state = 7;
+		repeated provisioner.GitAuthProvider git_auth_providers = 6;
```
Comment (deansheather) [inline provisionerd/proto/provisionerd.proto:20]: these two lines have tabs instead of 8 spaces

**Review by deansheather — APPROVED** [review]

File: coderd/provisionerdserver/provisionerdserver.go (hunk 1 above)
Comment (kylecarbs) [reply coderd/provisionerdserver/provisionerdserver.go:220]: I think we might make this an option in the data source eventually, but the default should be to never fail builds.

This feature is mostly useful for a first-time clone anyways.

**Pull request merged**